docs/design/2019-04-11-indexmerge.md

```diff
@@ -0,0 +1,284 @@
+# Proposal: scan a table using IndexMerge
+- Author(s) : WHU
+- Last updated : April 11
+- Discussion at :
+
+
+## 1 Abstract
+
+The proposal proposes to use multiple indexes to scan a table if possible. In some cases, using multiple indexes will improve performance.
+
+## 2 Background
+
+In present TiDB, a SQL statement with conditions involving multiple indexed attributes only uses one of the conditions as the index filter to build access condition, while others are regarded as table filters. Firstly, use index scan (at most one index) to get handles (rowid in TiDB). Then use the handles to get rows and check whether the rows satisfy the conditions of table filters. Some relational databases implement a table access path using multiple indexes. In some cases, this way will improve performance.
+
+We take an example to explain it. We define the table schema as :
+
+```
+	CREATE TABLE t1 (a int, b int, c int);
+	CREATE INDEX t1a on t1(a);
+	CREATE INDEX t1b on t1(b);
+	CREATE INDEX t1c on t1(c);
+```
+And use a test SQL statement `SELECT * FROM t1 where a < 2 or b > 50`. Currently, TiDB does a table scan  and puts `a < 2 or b > 50` as a Selection on top of it. If the selectivity of `a < 2 ` and `b > 50` is low, a better approach would be using indexes on columns `a` and `b` to retrieve rows respectively, and applying a union operation on the result sets.
+
+
+## 3 Proposal
+In short, we need to consider access paths using multiple indexes.
+
+### 3.1 Planner
+We propose to add new `IndexMergeReader / PhysicalIndexMergeReader` and `IndexMergeLookUpReader / PhysicalIndexMergeLookUpReader` operators.
+
+Now we just consider the following two kinds of queries: 
+
+（1）Conditions in CNF, e.g, `select * from t1 where c1 and c2 and c3 and …`

```

Comment:
I prefer to not build index intersection paths, since it would make code complicated while the performance gain is not guaranteed.

Reply:
I do some test for it. Then I found it have a little performance gain.

Reply:
@zz-jason What's your opinion on this?

Reply:
@hailanwhu could you describe more detail on your benchmark tests? what the data distribution and performance gain?

Reply:
I add it in the new version of pr.

Reply:
@zz-jason benchmark details have been added in the proposal, PTAL